op-service/rethdb-reader/src/lib.rs

```diff
@@ -15,19 +16,76 @@ pub unsafe extern "C" fn rdb_read_receipts(
     block_hash: *const u8,
     block_hash_len: usize,
     db_path: *const c_char,
-) -> ReceiptsResult {
-    read_receipts_inner(block_hash, block_hash_len, db_path).unwrap_or(ReceiptsResult::fail())
+) -> *mut ReceiptsResult {
+    let res = read_receipts_inner(block_hash, block_hash_len, db_path);
+    match res {
+        Ok(res) => Box::into_raw(Box::new(res)),
+        Err(err) => Box::into_raw(Box::new(ReceiptsResult::fail(err.to_string()))),
+    }
 }
 
-/// Free a string that was allocated in Rust and passed to C.
+/// Free a ReceiptsResult that was allocated in Rust and passed to C.
 ///
 /// # Safety
 /// - All possible nil pointer dereferences are checked.
 #[no_mangle]
-pub unsafe extern "C" fn rdb_free_string(string: *mut c_char) {
-    // Convert the raw pointer back to a CString and let it go out of scope,
+pub unsafe extern "C" fn rdb_free_receipts(res: *mut ReceiptsResult) {
+    // Convert the raw pointer back to a ReceiptResult and let it go out of scope,
     // which will deallocate the memory.
-    if !string.is_null() {
-        let _ = std::ffi::CString::from_raw(string);
+    if !res.is_null() {
+        if !(*res).data.is_null() {
+            // Same deal with the data string
+            let _ = CString::from_raw((*res).data as *mut c_char);
+        }
+        if !(*res).error.is_null() {
+            // Same deal with the error string
+            let _ = CString::from_raw((*res).error);
+        }
+        let _ = Box::from_raw(res);
+    }
+}
+
+/// Return the data from a ReceiptsResult.
+///
+/// # Safety
+/// - All possible nil pointer dereferences are checked.
+#[no_mangle]
+pub unsafe extern "C" fn rdb_get_receipts_data(res: *mut ReceiptsResult) -> *const c_char {
+    if res.is_null() {
+        return std::ptr::null();
+    }
+    let res = &*res;
+    if res.error.is_null() {
+        return std::ptr::null();
+    }
+    res.data as *const c_char
+}
+
+/// Return the length of the data from a ReceiptsResult.
+///
+/// # Safety
+/// - All possible nil pointer dereferences are checked.
+#[no_mangle]
+pub unsafe extern "C" fn rdb_get_receipts_data_len(res: *mut ReceiptsResult) -> usize {
+    if res.is_null() {
+        return 0;
+    }
+    let res = &*res;
+    if res.error.is_null() {
+        return 0;
+    }
+    res.data_len
+}
+
+/// Return the error string, if it exists, from a ReceiptsResult.
+///
+/// # Safety
+/// - All possible nil pointer dereferences are checked.
+#[no_mangle]
+pub unsafe extern "C" fn rdb_get_receipts_error(res: *mut ReceiptsResult) -> *const c_char {
+    if res.is_null() {
+        return std::ptr::null();
     }
+    let res = &*res;
+    res.error as *const c_char
```

Comment:
Doesn't look to provide any explicit guarantees that `res.error` is non-null

```suggestion
    res.as_ref().map_or_else(std::ptr::null, |res| {
        res.error
            .as_ref()
            .map_or_else(std::ptr::null, |e| e as *const c_char)
    })
```